pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed

pkg/ccl/testccl/workload/schemachange/schemachange_test.TestWorkload [failed](https://mesolite.cluster.engflow.com/invocations/default/52c9a036-2181-4b3f-82e5-e50f1352ccab?testReportRun=3&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL3Rlc3RjY2wvd29ya2xvYWQvc2NoZW1hY2hhbmdlOnNjaGVtYWNoYW5nZV90ZXN0) on master @ [99180451dea5c745ea9e5c86e60aeb125491be4d](https://github.com/cockroachdb/cockroach/commits/99180451dea5c745ea9e5c86e60aeb125491be4d):


```
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*operationGenerator).randOp
        	            	  | 	pkg/workload/schemachange/operation_generator.go:220
        	            	  | [...repeated from below...]
        	            	Wraps: (6) failed generating operation: selectStmt
        	            	Wraps: (7) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*operationGenerator).selectStmt
        	            	  | 	pkg/workload/schemachange/operation_generator.go:4913
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*operationGenerator).randOp
        	            	  | 	pkg/workload/schemachange/operation_generator.go:207
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).runInTxn
        	            	  | 	pkg/workload/schemachange/schemachange.go:591
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).run
        	            	  | 	pkg/workload/schemachange/schemachange.go:704
        	            	  | pkg/ccl/testccl/workload/schemachange/schemachange_test_test.TestWorkload.TestWorkload.func5.func7
        	            	  | 	pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:154
        	            	  | golang.org/x/sync/errgroup.(*Group).Go.func1
        	            	  | 	external/org_golang_x_sync/errgroup/errgroup.go:93
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1693
        	            	Wraps: (8) error getting random table name
        	            	Wraps: (9) ERROR: index-id "38" does not exist (SQLSTATE XXUUU)
        	            	Error types: (1) *markers.withMark (2) *schemachange.ErrorState (3) *withstack.withStack (4) *errutil.withPrefix (5) *withstack.withStack (6) *errutil.withPrefix (7) *withstack.withStack (8) *errutil.withPrefix (9) *pgconn.PgError
        	Test:       	TestWorkload
{
 "message": "Inspect Job Errors",
 "jobs": [
  {
   "jobId": 1133384032425213953,
   "status": "no errors reported"
  },
  {
   "jobId": 1133383984297213953,
   "status": "no errors reported"
  }
 ]
}
Schema Workload Stats
Total Schema Statements Executed = 103
Total Schema Statements Succeeded = 103
Total Schema Statement Expected Failures = 0
Total Transactions Committed = 58
Total Transactions Rolled Back = 60
Total Transactions Executed = 118
    schema_change_external_test.go:118: backup, tracing data, and system table dumps in /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/ccl/testccl/workload/schemachange/schemachange_test/run_3_of_10/test.outputs/logTestWorkload3913812892
    schema_change_external_test.go:127: running DROP with use_declarative_schema_changer = off
    panic.go:615: -- test log scope end --
test logs left over in: outputs.zip/logTestWorkload3913812892
--- FAIL: TestWorkload (151.23s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=3</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWorkload.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-57974

--- a/pkg/sql/catalog/lease/descriptor_state.go
+++ b/pkg/sql/catalog/lease/descriptor_state.go
@@ -47,6 +47,9 @@ type descriptorState struct {
 		// offline temporarily (as opposed to dropped).
 		takenOffline bool
 
+		// Timestamp at which this descriptor became offline.
+		takenOfflineAt hlc.Timestamp
+
 		// maxVersionSeen is used to prevent a race where a concurrent lease
 		// acquisition might miss an event indicating that there is a new version
 		// of a descriptor.
@@ -93,6 +96,12 @@ func (t *descriptorState) findForTimestamp(
 
 	// Acquire a lease if no descriptor exists in the cache.
 	if len(t.mu.active.data) == 0 {
+		// If the descriptor is marked as offline, we should attempt
+		// a historical query to fetch it, since it's likely querying
+		// the past. A non-historical query will fail instantly.
+		if t.mu.takenOffline {
+			return nil, false, errReadOlderVersion
+		}
 		return nil, false, errRenewLease
 	}
 	return t.findForTimestampImpl(ctx, timestamp.GetTimestamp(), timestamp.GetBaseTimestamp(), expensiveLogEnabled)
@@ -116,6 +125,12 @@ func (t *descriptorState) findForTimestampImpl(
 		// Check to see if the ModificationTime is valid. If only the initial version
 		// of the descriptor is known, then read it at the base timestamp.
 		if desc := t.mu.active.data[i]; desc.GetModificationTime().LessEq(leaseTimestamp) {
+			// If the version within the time range is marked as dropped,
+			// then return a dropped descriptor error.
+			if err := catalog.FilterDroppedDescriptor(desc); err != nil {
+				return nil, false, err
+			}
+
 			latest := i+1 == len(t.mu.active.data)
 			if !desc.hasExpired(ctx, readTimestamp) {
 				// Existing valid descriptor version.
@@ -130,7 +145,7 @@ func (t *descriptorState) findForTimestampImpl(
 				return t.findForTimestampImpl(ctx, readTimestamp, readTimestamp, expensiveLogEnabled)
 			}
 
-			if latest {
+			if latest && !t.mu.takenOffline {
 				// Renew the lease if the lease has expired
 				// The latest descriptor always has a lease.
 				return nil, false, errRenewLease
@@ -348,4 +363,19 @@ func (t *descriptorState) markAcquisitionDone(ctx context.Context) {
 	t.mu.Lock()
 	defer t.mu.Unlock()
 	t.mu.acquisitionsInProgress--
+}
+
+// setTakenOfflineLocked marks the descriptor as offline and sets / clears the timestamp.
+func (t *descriptorState) setTakenOfflineLocked(offline bool) {
+	// If the descriptor is already offline or online
+	// don't modify the timestamp.
+	if t.mu.takenOffline == offline {
+		return
+	}
+	t.mu.takenOffline = offline
+	timestamp := hlc.Timestamp{}
+	if offline {
+		timestamp = t.m.storage.clock.Now()
+	}
+	t.mu.takenOfflineAt = timestamp
 }

--- a/pkg/sql/catalog/lease/lease.go
+++ b/pkg/sql/catalog/lease/lease.go
@@ -754,6 +754,7 @@ func getDescriptorsFromStoreForInterval(
 	codec keys.SQLCodec,
 	id descpb.ID,
 	lowerBound, upperBound hlc.Timestamp,
+	isOffline bool,
 ) ([]historicalDescriptor, error) {
 	// Ensure lower bound is not an empty timestamp (now).
 	if lowerBound.IsEmpty() {
@@ -837,7 +838,12 @@ func getDescriptorsFromStoreForInterval(
 					if err != nil {
 						return err
 					}
-					if descContent == nil {
+					if len(descContent) == 0 {
+						// Skip any deletions of the descriptor.
+						if isOffline {
+							subsequentModificationTime = k.Timestamp
+							continue
+						}
 						return errors.Wrapf(errors.New("unsafe value error"), "error "+
 							"extracting raw bytes of descriptor with key %s modified between "+
 							"%s, %s", k.String(), k.Timestamp, subsequentModificationTime)
@@ -907,13 +913,21 @@ func (m *Manager) readOlderVersionForTimestamp(
 	if t == nil {
 		return nil, nil
 	}
-	endTimestamp, done := func() (hlc.Timestamp, bool) {
+	endTimestamp, isOffline, done := func() (hlc.Timestamp, bool, bool) {
 		t.mu.Lock()
 		defer t.mu.Unlock()
 
 		// If there are no descriptors, then we won't have a valid end timestamp.
 		if len(t.mu.active.data) == 0 {
-			return hlc.Timestamp{}, true
+			// If the descriptor is offline, then we can return when it was taken
+			// offline, a valid descriptor exists within this interval.
+			// Note: This allows us to populate dropped descriptors, which will
+			// be cleaned up via purgeOldVersions when these descriptors are actually
+			// deleted from the descriptor table (via the range feed).
+			if t.mu.takenOffline {
+				return t.mu.takenOfflineAt, true, false
+			}
+			return hlc.Timestamp{}, false, true
 		}
 		// We permit gaps in historical versions. We want to find the timestamp
 		// that represents the start of the validity interval for the known version
@@ -935,9 +949,9 @@ func (m *Manager) readOlderVersionForTimestamp(
 			// whether the descriptor for which we're searching actually exists. This
 			// will deal with cases where a concurrent fetch filled it in for us.
 			i > 0 && timestamp.Less(t.mu.active.data[i-1].getExpiration(ctx)) {
-			return hlc.Timestamp{}, true
+			return hlc.Timestamp{}, false, true
 		}
-		return t.mu.active.data[i].GetModificationTime(), false
+		return t.mu.active.data[i].GetModificationTime(), false, false
 	}()
 	if done {
 		return nil, nil
@@ -946,7 +960,7 @@ func (m *Manager) readOlderVersionForTimestamp(
 	// Retrieve descriptors in range [timestamp, endTimestamp) in decreasing
 	// modification time order.
 	descs, err := getDescriptorsFromStoreForInterval(
-		ctx, m.storage.db.KV(), m.Codec(), id, timestamp, endTimestamp,
+		ctx, m.storage.db.KV(), m.Codec(), id, timestamp, endTimestamp, isOffline,
 	)
 	if err != nil {
 		return nil, err
@@ -965,8 +979,9 @@ func (m *Manager) readOlderVersionForTimestamp(
 
 	// Unless the timestamp is exactly at the earliest modification time from
 	// ExportRequest, we'll invoke another call to retrieve the descriptor with
-	// modification time prior to the timestamp.
-	if timestamp.Less(earliestModificationTime) {
+	// modification time prior to the timestamp. For offline descriptors, we will
+	// try reading again if nothing was read above as a last resort.
+	if timestamp.Less(earliestModificationTime) || (isOffline && len(descs) == 0) {
 		desc, err := m.storage.getForExpiration(ctx, earliestModificationTime, id)
 		if err != nil {
 			return nil, err
@@ -1084,8 +1099,8 @@ func (m *Manager) upsertDescriptorIntoState(
 		return errors.AssertionFailedf("could not find descriptor state for id %d", id)
 	}
 	t.mu.Lock()
-	t.mu.takenOffline = false
 	defer t.mu.Unlock()
+	t.setTakenOfflineLocked(false)
 	err := t.upsertLeaseLocked(ctx, desc, session, m.storage.getRegionPrefix())
 	if err != nil {
 		return err
@@ -1420,7 +1435,7 @@ func (m *Manager) purgeOldVersions(
 		leases, leaseToExpire := func() (leasesToRemove []*storedLease, leasesToExpire *descriptorVersionState) {
 			t.mu.Lock()
 			defer t.mu.Unlock()
-			t.mu.takenOffline = dropped
+			t.setTakenOfflineLocked(dropped)
 			return t.removeInactiveVersions(ctx), t.mu.active.findPreviousToExpire(dropped)
 		}()
 		for _, l := range leases {

--- a/pkg/sql/catalog/lease/lease_internal_test.go
+++ b/pkg/sql/catalog/lease/lease_internal_test.go
@@ -1564,7 +1564,7 @@ func TestGetDescriptorsFromStoreForIntervalCPULimiterPagination(t *testing.T) {
 	var tableID int
 	sqlDB.QueryRow(t, `SELECT id FROM system.namespace WHERE name = 'baz'`).Scan(&tableID)
 	descs, err := getDescriptorsFromStoreForInterval(ctx, kvDB, s.Codec(), descpb.ID(tableID),
-		beforeCreate, afterCreate)
+		beforeCreate, afterCreate, false /*isOffline */)
 	require.NoError(t, err)
 	require.Len(t, descs, 3)
 	require.Equal(t, numRequests, 1)

--- a/pkg/sql/catalog/lease/lease_test.go
+++ b/pkg/sql/catalog/lease/lease_test.go
@@ -718,6 +718,7 @@ func TestLeasesOnDeletedTableAreReleasedImmediately(t *testing.T) {
 	srv, db, kvDB := serverutils.StartServer(t, params)
 	defer srv.Stopper().Stop(context.Background())
 	s := srv.ApplicationLayer()
+	lm := s.LeaseManager().(*lease.Manager)
 
 	_, err := db.Exec(`SET CLUSTER SETTING sql.defaults.use_declarative_schema_changer = 'off'`)
 	if err != nil {
@@ -759,6 +760,7 @@ CREATE TABLE test.t(a INT PRIMARY KEY);
 	mu.Unlock()
 
 	// DROP the table
+	ts := s.Clock().Now()
 	_, err = db.Exec(`DROP TABLE test.t`)
 	if err != nil {
 		t.Fatal(err)
@@ -783,6 +785,12 @@ CREATE TABLE test.t(a INT PRIMARY KEY);
 	if !testutils.IsError(err, "descriptor is being dropped") {
 		t.Fatalf("got a different error than expected: %v", err)
 	}
+	// Validate we can read the descriptor before the drop.
+	pastLease, err := lm.Acquire(ctx, lease.TimestampToReadTimestamp(ts), tableDesc.GetID())
+	require.NoError(t, err)
+	require.Equalf(t, lease3.Underlying().GetVersion(), pastLease.Underlying().GetVersion(), "prior version does not match ")
+	pastLease.Release(ctx)
+
 }
 
 // TestSubqueryLeases tests that all leases acquired by a subquery are

--- a/pkg/sql/show_create_clauses.go
+++ b/pkg/sql/show_create_clauses.go
@@ -552,11 +552,12 @@ func showComments(
 	}
 
 	for _, indexComment := range tc.indexes {
-		idx, err := catalog.MustFindIndexByID(table, descpb.IndexID(indexComment.subID))
-		if err != nil {
-			return err
+		// With leased descriptors enabled, it is possible that the descriptor
+		// we are currently using may not have these comments apply yet.
+		idx := catalog.FindIndexByID(table, descpb.IndexID(indexComment.subID))
+		if idx == nil {
+			continue
 		}
-
 		f.WriteString(";\n")
 		f.FormatNode(&tree.CommentOnIndex{
 			Index: tree.TableIndexName{
@@ -568,11 +569,13 @@ func showComments(
 	}
 
 	for _, constraintComment := range tc.constraints {
-		f.WriteString(";\n")
-		c, err := catalog.MustFindConstraintByID(table, descpb.ConstraintID(constraintComment.subID))
-		if err != nil {
-			return err
+		// With leased descriptors enabled, it is possible that the descriptor
+		// we are currently using may not have these constraint apply yet.
+		c := catalog.FindConstraintByID(table, descpb.ConstraintID(constraintComment.subID))
+		if c == nil {
+			continue
 		}
+		f.WriteString(";\n")
 		f.FormatNode(&tree.CommentOnConstraint{
 			Constraint: tree.Name(c.GetName()),
 			Table:      tn.ToUnresolvedObjectName(),

